aten/src/ATen/cuda/Atomic.cuh

```diff
@@ -35,6 +35,26 @@ struct AtomicFPOp<at::Half> {
   }
 };
 
+template <>
+struct AtomicFPOp<c10::complex<float>> {
+  template <typename func_t>
+  inline __device__ c10::complex<float> operator() (c10::complex<float> *address, c10::complex<float> val, const func_t& func) {
+    unsigned long long int* addr_as_ull = (unsigned long long int*)address;
+    unsigned long long int old = *addr_as_ull;
+    unsigned long long int assumed, new_val;
+
+    c10::complex<float> csum;
+    do {
+        assumed = old;
+        csum = func(csum, val);
+        new_val = *reinterpret_cast<unsigned long long*>(&csum);
+        old = atomicCAS(addr_as_ull, assumed, new_val);
+    } while (assumed != old);
+
+    return *reinterpret_cast<c10::complex<float>*>(&addr_as_ull);
```

Comment:
This isn't atomic? You need to return `csum` directly, otherwise the value at `addr_as_ull` may change underneath you.

Reply:
In fact this is also wrong as atomic read-modify-write ops return the old value, not the new value. So this should be bit-casting `assumed`.

Reply:
> In fact this is also wrong as atomic read-modify-write ops return the old value, not the new value. So this should be bit-casting `assumed`.

Sorry for the oversight. Could you help me understand? I know that `atomicCAS` returns the old value but with what in mind are you referring that to?

Reply:
`assumed` is the value before performing the update, which is what is returned by normal `atomicAdd`, `atomicMax`, etc.

See the CAS implementation for half as an example:
https://github.com/pytorch/pytorch/blob/6bdcc8d58867ed7a87e53939d502f667a9869bf2/aten/src/ATen/cuda/Atomic.cuh#L33-L34

Reply:
Oh you're right. I got tunnel visioned on the line on the actual call of `atomicCAS`, yes it should be `assumed`.